src/main.js

```diff
@@ -0,0 +1,296 @@
+      return this._request(extendedOptions)
+    }
+  }
+
+  _generateResourceMethod (operation, query = '') {
+    return (...args) => {
+      // by using ...args we get an array of arguments
+      // similar to the arguments object in non-arrow functions
+
+      // the arguments for a resource method are:
+      // client.<resource method>(path, [data], requestOptions)
+
+      // the data argument is not exactly optional, it will not exist
+      // for all methods except the ones using 'PUT', where it will
+      // have a fixed second spot (and requestOptions becomes third)
+      const path = args[0]
+
+      // merge all request options
+      const requestOptions = Object.assign({},
+        this._getOption('requestOptions'),
+        operation === 'PUT' ? args[2] : args[1])
+
+      // generate request options
+      requestOptions.path = `/${path}${query}`
+      requestOptions.method = operation
+      if (operation === 'PUT') requestOptions.json = args[1]
+
+      // execute request
+      return this._request(requestOptions)
+    }
+  }
+
+  // ----------------
+  // requests
+
+  _extendRequestOptions (original, options) {
+    const schema = original.schema.query
+    // no schema for the query -> no need to extend
+    if (!schema) return options
+    // create params array
+    const params = []
+    // loop ONLY through schema params
+    for (const key of Object.keys(schema.properties)) {
+      // if and only if the property exists in the new
+      // params, add it to the extended params array
+      key in options.json &&
+        params.push(`${key}=${encodeURIComponent(options.json[key])}`)
+    }
+    // this way, only whitelisted params (on the schema)
+    // will be allowed in the request
+
+    // add params (if any) to path as URL query
+    if (params.length) options.path += `?${params.join('&')}`
+    return options
+  }
+
+  _formatRequest (uriTemplate, values = {}) {
+    return uriTemplate.replace(/{{([\S]+?)}}/g, (a, b, c) => {
+      const parts = b.split('=')
+      const prop = parts[0]
+      const defaultValue = parts[1] || ''
+      return values[prop] || defaultValue
+    })
+  }
+
+  async _request (options = {}) {
+    // validate
+    if (!tv4.validate(options, REQUEST_SCHEMA)) throw tv4.error
+    // create URI template
+    const uriTemplate = `${this._getOption('endpoint')}/${this._getOption('apiVersion')}${options.path}`
+    // replace variables in uri
+    const uri = this._formatRequest(uriTemplate, options.json)
+      // replace unicode encodings
+      .replace(/&#x2F;/g, '/')
+    // add URI
+    options.uri = uri
+    // add headers
+    options.headers = options.headers || {}
+    // add token
+    const token = this._getOption('token')
+    typeof token === 'string' && token.length &&
+      (options.headers['X-Vault-Token'] = token)
+
+    // execute request
+    debug(options.method, uri)
+    const response = await this.request(options)
```

Comment:
This is the only place where we use `await` sure to drop Node 6-7 ? We could archive it with promises only.

Reply:
If we want compatibility (which we probably do) then we should use babel. I don't like using outdated or old code and design patterns when you have the newest functionality live on a LTS node version.

Reply:
Much more people think this about transpiling code 😆 I think ES7 features aren't needed for a http client.

Reply:
[Added support for node v6+](https://github.com/DaniGuardiola/node-vault/commit/1836216df2514801a6755b466e52b0d0194ccb34) ([travis support](https://github.com/kr1sp1n/node-vault/commit/bdbeff1e06ece732e04c0cb2f2db205a54a19819))

Without babel :)